In [1]:
import pandas as pd 
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
sns.set()

In [2]:
hr=pd.read_csv("D:\\LetsUpgrade\\Day 7\\general_data.csv")

In [3]:
hr.head()

Age Attrition     BusinessTravel              Department  DistanceFromHome  \
0   51        No      Travel_Rarely                   Sales                 6   
1   31       Yes  Travel_Frequently  Research & Development                10   
2   32        No  Travel_Frequently  Research & Development                17   
3   38        No         Non-Travel  Research & Development                 2   
4   32        No      Travel_Rarely  Research & Development                10   

   Education EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0          2  Life Sciences              1           1  Female  ...   
1          1  Life Sciences              1           2  Female  ...   
2          4          Other              1           3    Male  ...   
3          5  Life Sciences              1           4    Male  ...   
4          1        Medical              1           5    Male  ...   

   NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0                 1.0      Y                11              8   
1                 0.0      Y                23              8   
2                 1.0      Y                15              8   
3                 3.0      Y                11              8   
4                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [4]:
data_Yes=hr[hr['Attrition']=='Yes']
data_No=hr[hr['Attrition']=='No']

In [5]:
data_Yes.head()

Age Attrition     BusinessTravel              Department  \
1    31       Yes  Travel_Frequently  Research & Development   
6    28       Yes      Travel_Rarely  Research & Development   
13   47       Yes         Non-Travel  Research & Development   
28   44       Yes  Travel_Frequently  Research & Development   
30   26       Yes      Travel_Rarely  Research & Development   

    DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                 10          1  Life Sciences              1           2   
6                 11          2        Medical              1           7   
13                 1          1        Medical              1          14   
28                 1          2        Medical              1          29   
30                 4          3        Medical              1          31   

    Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1   Female  ...                 0.0      Y                23              8   
6     Male  ...                 2.0      Y                20              8   
13    Male  ...                 1.0      Y                11              8   
28    Male  ...                 3.0      Y                14              8   
30    Male  ...                 2.0      Y                11              8   

    StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
1                  1               6.0                      3               5   
6                  1               5.0                      2               0   
13                 2              10.0                      4              10   
28                 1              19.0                      2               1   
30                 0               5.0                      5               3   

    YearsSinceLastPromotion  YearsWithCurrManager  
1                         1                     4  
6                         0                     0  
13                        9                     9  
28                        0                     0  
30                        0                     2  

[5 rows x 24 columns]

In [6]:
data_Yes.isnull()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  \
1     False      False           False       False             False   
6     False      False           False       False             False   
13    False      False           False       False             False   
28    False      False           False       False             False   
30    False      False           False       False             False   
...     ...        ...             ...         ...               ...   
4381  False      False           False       False             False   
4386  False      False           False       False             False   
4388  False      False           False       False             False   
4391  False      False           False       False             False   
4402  False      False           False       False             False   

      Education  EducationField  EmployeeCount  EmployeeID  Gender  ...  \
1         False           False          False       False   False  ...   
6         False           False          False       False   False  ...   
13        False           False          False       False   False  ...   
28        False           False          False       False   False  ...   
30        False           False          False       False   False  ...   
...         ...             ...            ...         ...     ...  ...   
4381      False           False          False       False   False  ...   
4386      False           False          False       False   False  ...   
4388      False           False          False       False   False  ...   
4391      False           False          False       False   False  ...   
4402      False           False          False       False   False  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
1                  False   False              False          False   
6                  False   False              False          False   
13                 False   False              False          False   
28                 False   False              False          False   
30                 False   False              False          False   
...                  ...     ...                ...            ...   
4381               False   False              False          False   
4386               False   False              False          False   
4388               False   False              False          False   
4391               False   False              False          False   
4402               False   False              False          False   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
1                False              False                  False   
6                False              False                  False   
13               False              False                  False   
28               False              False                  False   
30               False              False                  False   
...                ...                ...                    ...   
4381             False              False                  False   
4386             False              False                  False   
4388             False              False                  False   
4391             False              False                  False   
4402             False              False                  False   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1              False                    False                 False  
6              False                    False                 False  
13             False                    False                 False  
28             False                    False                 False  
30             False                    False                 False  
...              ...                      ...                   ...  
4381           False                    False                 False  
4386           False    

In [7]:
data_Yes.dropna()

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                   10          1  Life Sciences              1           2   
6                   11          2        Medical              1           7   
13                   1          1        Medical              1          14   
28                   1          2        Medical              1          29   
30                   4          3        Medical              1          31   
...                ...        ...            ...            ...         ...   
4381                 7          1  Life Sciences              1        4382   
4386                11          4      Marketing              1        4387   
4388                 1          3  Life Sciences              1        4389   
4391                23          1  Life Sciences              1        4392   
4402                 2          3      Marketing              1        4403   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1     Female  ...                 0.0      Y                23              8   
6       Male  ...                 2.0      Y                20              8   
13      Male  ...                 1.0      Y                11              8   
28      Male  ...                 3.0      Y                14              8   
30      Male  ...                 2.0      Y                11              8   
...      ...  ...                 ...    ...               ...            ...   
4381  Female  ...                 1.0      Y                21              8   
4386    Male  ...                 5.0      Y                21              8   
4388    Male  ...                 7.0      Y                11              8   
4391    Male  ...                 0.0      Y                11              8   
4402    Male  ...                 6.0      Y                11              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
1                    1               6.0                      3   
6                    1               5.0                      2   
13                   2              10.0                      4   
28                   1              19.0                      2   
30                   0               5.0                      5   
...                ...               ...                    ...   
4381                 0               4.0                      2   
4386                 0               8.0                      2   
4388                 0              13.0                      2   
4391                 0               4.0                      2   
4402                 1              17.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1                  5                        1                     4  
6                  0                        0                     0  
13                10                        9                     9  
28                 1                        0                     0  
30                 3                        0                     2  
...              ...                      ...                   ..

In [8]:
data_No.isnull()

Age  Attrition  BusinessTravel  Department  DistanceFromHome  \
0     False      False           False       False             False   
2     False      False           False       False             False   
3     False      False           False       False             False   
4     False      False           False       False             False   
5     False      False           False       False             False   
...     ...        ...             ...         ...               ...   
4405  False      False           False       False             False   
4406  False      False           False       False             False   
4407  False      False           False       False             False   
4408  False      False           False       False             False   
4409  False      False           False       False             False   

      Education  EducationField  EmployeeCount  EmployeeID  Gender  ...  \
0         False           False          False       False   False  ...   
2         False           False          False       False   False  ...   
3         False           False          False       False   False  ...   
4         False           False          False       False   False  ...   
5         False           False          False       False   False  ...   
...         ...             ...            ...         ...     ...  ...   
4405      False           False          False       False   False  ...   
4406      False           False          False       False   False  ...   
4407      False           False          False       False   False  ...   
4408      False           False          False       False   False  ...   
4409      False           False          False       False   False  ...   

      NumCompaniesWorked  Over18  PercentSalaryHike  StandardHours  \
0                  False   False              False          False   
2                  False   False              False          False   
3                  False   False              False          False   
4                  False   False              False          False   
5                  False   False              False          False   
...                  ...     ...                ...            ...   
4405               False   False              False          False   
4406               False   False              False          False   
4407               False   False              False          False   
4408               False   False              False          False   
4409               False   False              False          False   

      StockOptionLevel  TotalWorkingYears  TrainingTimesLastYear  \
0                False              False                  False   
2                False              False                  False   
3                False              False                  False   
4                False              False                  False   
5                False              False                  False   
...                ...                ...                    ...   
4405             False              False                  False   
4406             False              False                  False   
4407             False              False                  False   
4408             False              False                  False   
4409             False               True                  False   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0              False                    False                 False  
2              False                    False                 False  
3              False                    False                 False  
4              False                    False                 False  
5              False                    False                 False  
...              ...                      ...                   ...  
4405           False                    False                 False  
4406           False    

In [9]:
data_No.dropna()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
5      46        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4404   29        No      Travel_Rarely                   Sales   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
5                    8          3  Life Sciences              1           6   
...                ...        ...            ...            ...         ...   
4404                 4          3          Other              1        4405   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
5     Female  ...                 3.0      Y                13              8   
...      ...  ...                 ...    ...               ...            ...   
4404  Female  ...                 1.0      Y                18              8   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
5                    0              28.0                      5   
...                ...               ...                    ...   
4404                 0               6.0                      2   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
5                  7                        7                     7  
...              ...                      ...                   ..

In [10]:
data_Yes.duplicated()

1       False
6       False
13      False
28      False
30      False
        ...  
4381    False
4386    False
4388    False
4391    False
4402    False
Length: 711, dtype: bool

In [11]:
data_Yes.drop_duplicates()

Age Attrition     BusinessTravel              Department  \
1      31       Yes  Travel_Frequently  Research & Development   
6      28       Yes      Travel_Rarely  Research & Development   
13     47       Yes         Non-Travel  Research & Development   
28     44       Yes  Travel_Frequently  Research & Development   
30     26       Yes      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4381   29       Yes      Travel_Rarely  Research & Development   
4386   33       Yes      Travel_Rarely                   Sales   
4388   33       Yes      Travel_Rarely                   Sales   
4391   32       Yes      Travel_Rarely                   Sales   
4402   37       Yes  Travel_Frequently                   Sales   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
1                   10          1  Life Sciences              1           2   
6                   11          2        Medical              1           7   
13                   1          1        Medical              1          14   
28                   1          2        Medical              1          29   
30                   4          3        Medical              1          31   
...                ...        ...            ...            ...         ...   
4381                 7          1  Life Sciences              1        4382   
4386                11          4      Marketing              1        4387   
4388                 1          3  Life Sciences              1        4389   
4391                23          1  Life Sciences              1        4392   
4402                 2          3      Marketing              1        4403   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
1     Female  ...                 0.0      Y                23              8   
6       Male  ...                 2.0      Y                20              8   
13      Male  ...                 1.0      Y                11              8   
28      Male  ...                 3.0      Y                14              8   
30      Male  ...                 2.0      Y                11              8   
...      ...  ...                 ...    ...               ...            ...   
4381  Female  ...                 1.0      Y                21              8   
4386    Male  ...                 5.0      Y                21              8   
4388    Male  ...                 7.0      Y                11              8   
4391    Male  ...                 0.0      Y                11              8   
4402    Male  ...                 6.0      Y                11              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
1                    1               6.0                      3   
6                    1               5.0                      2   
13                   2              10.0                      4   
28                   1              19.0                      2   
30                   0               5.0                      5   
...                ...               ...                    ...   
4381                 0               4.0                      2   
4386                 0               8.0                      2   
4388                 0              13.0                      2   
4391                 0               4.0                      2   
4402                 1              17.0                      2   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
1                  5                        1                     4  
6                  0                        0                     0  
13                10                        9                     9  
28                 1                        0                     0  
30                 3                        0                     2  
...              ...                      ...                   ..

In [12]:
data_No.duplicated()

0       False
2       False
3       False
4       False
5       False
        ...  
4405    False
4406    False
4407    False
4408    False
4409    False
Length: 3699, dtype: bool

In [13]:
data_No.drop_duplicates()

Age Attrition     BusinessTravel              Department  \
0      51        No      Travel_Rarely                   Sales   
2      32        No  Travel_Frequently  Research & Development   
3      38        No         Non-Travel  Research & Development   
4      32        No      Travel_Rarely  Research & Development   
5      46        No      Travel_Rarely  Research & Development   
...   ...       ...                ...                     ...   
4405   42        No      Travel_Rarely  Research & Development   
4406   29        No      Travel_Rarely  Research & Development   
4407   25        No      Travel_Rarely  Research & Development   
4408   42        No      Travel_Rarely                   Sales   
4409   40        No      Travel_Rarely  Research & Development   

      DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                    6          2  Life Sciences              1           1   
2                   17          4          Other              1           3   
3                    2          5  Life Sciences              1           4   
4                   10          1        Medical              1           5   
5                    8          3  Life Sciences              1           6   
...                ...        ...            ...            ...         ...   
4405                 5          4        Medical              1        4406   
4406                 2          4        Medical              1        4407   
4407                25          2  Life Sciences              1        4408   
4408                18          2        Medical              1        4409   
4409                28          3        Medical              1        4410   

      Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0     Female  ...                 1.0      Y                11              8   
2       Male  ...                 1.0      Y                15              8   
3       Male  ...                 3.0      Y                11              8   
4       Male  ...                 4.0      Y                12              8   
5     Female  ...                 3.0      Y                13              8   
...      ...  ...                 ...    ...               ...            ...   
4405  Female  ...                 3.0      Y                17              8   
4406    Male  ...                 2.0      Y                15              8   
4407    Male  ...                 0.0      Y                20              8   
4408    Male  ...                 0.0      Y                14              8   
4409    Male  ...                 0.0      Y                12              8   

      StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  \
0                    0               1.0                      6   
2                    3               5.0                      2   
3                    3              13.0                      5   
4                    2               9.0                      2   
5                    0              28.0                      5   
...                ...               ...                    ...   
4405                 1              10.0                      5   
4406                 0              10.0                      2   
4407                 0               5.0                      4   
4408                 1              10.0                      2   
4409                 0               NaN                      6   

      YearsAtCompany  YearsSinceLastPromotion  YearsWithCurrManager  
0                  1                        0                     0  
2                  5                        0                     3  
3                  8                        7                     5  
4                  6                        0                     4  
5                  7                        7                     7  
...              ...                      ...                   ..

# Statistical Test

## 1. Mannwhitenyu Test: It is used to compare to independent sample

In [14]:
from scipy.stats import mannwhitneyu

In [15]:
stats,p=mannwhitneyu(data_Yes.DistanceFromHome,data_No.DistanceFromHome)
print(stats,p)

1312110.0 0.4629185205822659


___As p value is less than 0.05, thus we can conclude that alternative hypothesis is rejected i.e there is a no significant difference of the Distance from home between Attrition Yes and Attrition No___ 

In [16]:
stats,p=mannwhitneyu(data_Yes.MonthlyIncome,data_No.MonthlyIncome)
print(stats,p)

1264900.5 0.053577283839938566


___As p value is greater than 0.05, thus we can canclude that alternative hypothesis is accepted i.e there is a significant difference of the Monthly Income between Attrition Yes and Attrition No___

In [17]:
stats,p=mannwhitneyu(data_Yes.TotalWorkingYears,data_No.TotalWorkingYears)
print(stats,p)

909954.0 2.8897554531400447e-39


___As p value is less than 0.05, thus we can conclude that alternative hypothesis is rejected i.e there is a no significant difference TotalWorkingYears between Attrition Yes and Attrition No___ 

In [18]:
stats,p=mannwhitneyu(data_Yes.YearsAtCompany,data_No.YearsAtCompany)
print(stats,p)

923238.0 6.047598261692858e-37


___As p value is less than 0.05, thus we can conclude that alternative hypothesis is rejected i.e there is a no significant difference of YeasAtCompany between Attrition Yes and Attrition No___ 

## Chi Square T Test : It is used to check the dependency of the categorical variables 

In [19]:
from scipy.stats import chi2_contingency

In [20]:
chitable=pd.crosstab(hr.BusinessTravel,hr.Over18)

In [21]:
stats,p,dof,expected=chi2_contingency(chitable)
print(stats,p)

0.0 1.0


___As p value is greater than 0.05, thus we can conclude that alternative hypothesis is acseptedi.e there is a dependance between BusinessTravel and Over18___ 

In [22]:
hr.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'Department', 'DistanceFromHome',
       'Education', 'EducationField', 'EmployeeCount', 'EmployeeID', 'Gender',
       'JobLevel', 'JobRole', 'MaritalStatus', 'MonthlyIncome',
       'NumCompaniesWorked', 'Over18', 'PercentSalaryHike', 'StandardHours',
       'StockOptionLevel', 'TotalWorkingYears', 'TrainingTimesLastYear',
       'YearsAtCompany', 'YearsSinceLastPromotion', 'YearsWithCurrManager'],
      dtype='object')

In [23]:
chitable=pd.crosstab(hr.Department,hr.EducationField)

In [24]:
stats,p,dof,expected=chi2_contingency(chitable)
print(stats,p)

3074.937741992298 0.0


___As p value is less than 0.05, thus we can conclude that alternative hypothesis is rejected i.e there is a no dependance between Departement and EducationField___ 

In [25]:
chitable=pd.crosstab(hr.Gender,hr.JobLevel)

In [26]:
stats,p,dof,expected=chi2_contingency(chitable)
print(stats,p)

10.290012909944652 0.03581578139884318


___As p value is less than 0.05, thus we can conclude that alternative hypothesis is rejected i.e there is a no dependance between Gender and Job role___ 

## 3.Kruskal wallis Test: It is used to compare the more than two independent sample

In [27]:
from scipy.stats import kruskal

In [28]:
stats,p=kruskal(hr.YearsAtCompany,hr.NumCompaniesWorked,hr.MonthlyIncome)
print(stats,p)

nan nan


## 4.One way anova: It is used when dependent variable is continuos and indepent variable is categorical

In [29]:
from sklearn import preprocessing 
Label_encoder=preprocessing.LabelBinarizer()

In [30]:
hr['Attrition'] = Label_encoder.fit_transform(hr['Attrition']) 

In [31]:
hr.head()

Age  Attrition     BusinessTravel              Department  \
0   51          0      Travel_Rarely                   Sales   
1   31          1  Travel_Frequently  Research & Development   
2   32          0  Travel_Frequently  Research & Development   
3   38          0         Non-Travel  Research & Development   
4   32          0      Travel_Rarely  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeID  \
0                 6          2  Life Sciences              1           1   
1                10          1  Life Sciences              1           2   
2                17          4          Other              1           3   
3                 2          5  Life Sciences              1           4   
4                10          1        Medical              1           5   

   Gender  ...  NumCompaniesWorked Over18 PercentSalaryHike  StandardHours  \
0  Female  ...                 1.0      Y                11              8   
1  Female  ...                 0.0      Y                23              8   
2    Male  ...                 1.0      Y                15              8   
3    Male  ...                 3.0      Y                11              8   
4    Male  ...                 4.0      Y                12              8   

   StockOptionLevel TotalWorkingYears  TrainingTimesLastYear  YearsAtCompany  \
0                 0               1.0                      6               1   
1                 1               6.0                      3               5   
2                 3               5.0                      2               5   
3                 3              13.0                      5               8   
4                 2               9.0                      2               6   

   YearsSinceLastPromotion  YearsWithCurrManager  
0                        0                     0  
1                        1                     4  
2                        0                     3  
3                        7                     5  
4                        0                     4  

[5 rows x 24 columns]

In [32]:
import statsmodels.api as sm

In [33]:
from statsmodels.formula.api import ols

In [34]:
models=ols("Attrition~C(Gender)",hr).fit()

In [35]:
oneway=sm.stats.anova_lm(models,typ=2)
print(oneway)

               sum_sq      df         F   PR(>F)
C(Gender)    0.195918     1.0  1.448585  0.22882
Residual   596.173469  4408.0       NaN      NaN


___we can conclude that here between variation is 0.195918 and within variation is 596.1734 , here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no signifance differenace between attrition and gender___

In [36]:
models=ols("Attrition~C(BusinessTravel)",hr).fit()

In [37]:
oneway=sm.stats.anova_lm(models,typ=2)
print(oneway)

                       sum_sq      df          F        PR(>F)
C(BusinessTravel)    9.810647     2.0  36.855236  1.337694e-16
Residual           586.558741  4407.0        NaN           NaN


___Here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no signifance differenace between attrition and BussinessTravel___

In [38]:
models=ols("Attrition~C(Department)",hr).fit()

In [39]:
oneway=sm.stats.anova_lm(models,typ=2)
print(oneway)

                   sum_sq      df         F        PR(>F)
C(Department)    3.933911     2.0  14.63176  4.639936e-07
Residual       592.435476  4407.0       NaN           NaN


___Here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no signifance differenace between attrition and Departement___

In [40]:
models=ols("Attrition~C(EducationField)",hr).fit()

In [41]:
oneway=sm.stats.anova_lm(models,typ=2)
print(oneway)

                       sum_sq      df         F        PR(>F)
C(EducationField)    6.246992     5.0  9.324084  7.560112e-09
Residual           590.122395  4404.0       NaN           NaN


___Here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no signifance differenace between attrition and JobRole___

In [42]:
models=ols("Attrition~C(MaritalStatus)",hr).fit()

In [43]:
oneway=sm.stats.anova_lm(models,typ=2)
print(oneway)

                      sum_sq      df          F        PR(>F)
C(MaritalStatus)   18.728302     2.0  71.441963  2.920874e-31
Residual          577.641086  4407.0        NaN           NaN


___Here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no signifance differenace between attrition and Marital Status___

## Two way anova: one dependent variable and more than two independent variable

In [44]:
models=ols("Attrition~C(Gender)+C(BusinessTravel)",hr).fit()

In [45]:
twoway=sm.stats.anova_lm(models,typ=2)
print(twoway)

                       sum_sq      df          F        PR(>F)
C(Gender)            0.243508     1.0   1.829893  1.762096e-01
C(BusinessTravel)    9.858236     2.0  37.040986  1.114415e-16
Residual           586.315233  4406.0        NaN           NaN


___Here P value is less than 0.05 hence we can reject the null hypothesis i.e there is no variation differenace between attrition, Gender and Bussiness Travel___